miden-crypto/src/merkle/mmr/forest.rs

```diff
@@ -0,0 +1,245 @@
+    pub fn smallest_tree_checked(self) -> Forest {
+        if self.0 > 0 {
+            self.smallest_tree()
+        } else {
+            Forest::empty()
+        }
+    }
+
+    /// Keeps only trees larger than the reference tree.
+    pub fn trees_larger_than(self, base: u32) -> Forest {
+        self & high_bitmask(base + 1)
+    }
+
+    /// Creates a new forest with all possible trees smaller than
+    /// the smallest tree in this forest.
+    pub fn all_smaller_trees(self) -> Forest {
+        debug_assert!(self.0.count_ones() == 1);
+        Forest::with_leaves(self.0 - 1)
+    }
+
+    /// Returns true if the forest containts a single-node tree
+    pub fn has_odd_leaf(self) -> bool {
+        self.0 & 1 != 0
+    }
+
+    /// Add a single-node tree if not already present in the forest.
+    pub fn odd_leaf_added(self) -> Forest {
+        Forest::with_leaves(self.0 | 1)
+    }
+
+    /// Remove the single-node tree if present in the forest.
+    pub fn odd_leaf_removed(self) -> Forest {
+        Forest::with_leaves(self.0 & (usize::MAX << 1))
+    }
+
+    /// Given a 0-indexed leaf position in the current forest, return the tree number responsible
+    /// for the position.
+    ///
+    /// Note:
+    /// The result is a tree position `p`, it has the following interpretations. $p+1$ is the depth
+    /// of the tree. Because the root element is not part of the proof, $p$ is the length of the
+    /// authentication path. $2^p$ is equal to the number of leaves in this particular tree. and
+    /// $2^(p+1)-1$ corresponds to size of the tree.
+    pub fn leaf_to_corresponding_tree(self, pos: usize) -> Option<u32> {
+        let forest = self.0;
+
+        if pos >= forest {
+            None
+        } else {
+            // - each bit in the forest is a unique tree and the bit position its power-of-two size
+            // - each tree owns a consecutive range of positions equal to its size from
+            //   left-to-right
+            // - this means the first tree owns from `0` up to the `2^k_0` first positions, where
+            //   `k_0` is the highest true bit position, the second tree from `2^k_0 + 1` up to
+            //   `2^k_1` where `k_1` is the second highest bit, so on.
+            // - this means the highest bits work as a category marker, and the position is owned by
+            //   the first tree which doesn't share a high bit with the position
+            let before = forest & pos;
+            let after = forest ^ before;
+            let tree = after.ilog2();
+
+            Some(tree)
+        }
+    }
+
+    /// Given a 0-indexed leaf position in the current forest, return the 0-indexed leaf position
+    /// in the tree to which the leaf belongs.
+    pub fn leaf_relative_position(self, pos: usize) -> Option<usize> {
+        let tree = self.leaf_to_corresponding_tree(pos)?;
+        let forest_before = self & high_bitmask(tree + 1);
+        Some(pos - forest_before.0)
+    }
```

Comment:
Same here as above, `pos` -> `leaf_idx`, `tree` -> `tree_idx`.

I wonder if this method has to be public. I'd make it private (so `pub(super)` or `pub(crate)`) and only make it public if we notice we need it to be.